include/sys/metaslab_impl.h

```diff
@@ -129,6 +130,74 @@ typedef enum trace_alloc_type {
 #define	WEIGHT_GET_COUNT(weight)		BF64_GET((weight), 0, 55)
 #define	WEIGHT_SET_COUNT(weight, x)		BF64_SET((weight), 0, 55, x)
 
+#define	CAS_MAGIC		0xb10ca110ca7e57a7ULL
+#define	CAS_HISTOGRAM_SIZE	9
+
+/*
+ * Each alloction stat is comprised
+ * of a byte count and block count.
+ */
+struct alloc_stat {
```

Comment:
consider renaming all of these `alloc_stat` -> `class_stat`.

It's confusing that the "alloc stat bytes" is not ASIZE, it's PSIZE.  We might also consider if we want this to actually be ASIZE.  (But might need to make changes for gang blocks in that case - remember that GBH size is a constant, not BP_GET_ASIZE()).

I think as_blocks is actually the number of DVA's, not the number of logical blocks (blkptr_t's).

We should consider if we want to be tracking DVAs/allocations vs tracking logical blocks.  If we track logical blocks, we could call into the stat code at a higher level, and that might be more clean.  I think you wouldn't need to track what is a gang block child, since those are not logical blocks.  E.g. you could call into the stat code from the end of zio_dva_allocate(), and the beginning of metaslab_free().  Or maybe wherever you are doing INIT_BLOCK_CONTEXT() today?

Reply:
great points.  The code is now block-based and this simplified things and better aligns with what `zdb` -b is reporting.